File: 0200-number-of-islands/0200-number-of-islands.java
import java.util.*;

class Solution {
    public int numIslands(char[][] grid) {
        int output = 0;
        int m = grid.length;
        int n = grid[0].length;

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (grid[i][j] == '1') {
                    output++;
                    bfs(grid, i, j);
                }
            }
        }
        return output;
    }

    private void bfs(char[][] grid, int i, int j) {
        int m = grid.length;
        int n = grid[0].length;

        Queue<int[]> queue = new LinkedList<>();
        queue.offer(new int[]{i, j});
        grid[i][j] = '#';

        int[][] directions = {{-1,0},{1,0},{0,-1},{0,1}};

        while (!queue.isEmpty()) {
            int[] current = queue.poll();
            int row = current[0];
            int col = current[1];

            for (int[] dir : directions) {
                int newRow = row + dir[0];
                int newCol = col + dir[1];

                if (newRow >= 0 && newRow < m &&
                    newCol >= 0 && newCol < n &&
                    grid[newRow][newCol] == '1') {

                    queue.offer(new int[]{newRow, newCol});
                    grid[newRow][newCol] = '#';
                }
            }
        }
    }
}

File: 0200-number-of-islands/0200-number-of-islands.js
/**
 * @param {character[][]} grid
 * @return {number}
 */

var numIslands = function(grid) {
    const bArray = Array.from({ length: grid.length }, () => new Array(grid[0].length).fill(false));

    let output = 0;

    for (let i = 0; i < grid.length; i++) {
        for (let j = 0; j < grid[0].length; j++) {
            if (grid[i][j] == '1' && !bArray[i][j]) {
                output += 1;
                bfs(i, j, grid, bArray);
            }
        }
    }

    return output;
};

var bfs = function(i, j, grid, bArray) {
    if (i < 0 || i > grid.length - 1 ||
        j < 0 || j > grid[0].length - 1) return false;
    if (grid[i][j] === '0') return false;
    if (bArray[i][j]) return false;

    bArray[i][j] = true;

    bfs(i+1, j, grid, bArray);
    bfs(i-1, j, grid, bArray);
    bfs(i, j+1, grid, bArray);
    bfs(i, j-1, grid, bArray);
}